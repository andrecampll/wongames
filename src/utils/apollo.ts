import { useMemo } from 'react';
import { setContext } from '@apollo/client/link/context';
import {
  ApolloClient,
  HttpLink,
  NormalizedCache,
  NormalizedCacheObject,
} from '@apollo/client';
import { Session } from 'next-auth';
import apolloCache from './apolloCache';

let apolloClient: ApolloClient<NormalizedCacheObject | null>;

function createApolloClient(session?: Session | null) {
  const httpLink = new HttpLink({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
  });

  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || '';
    const authorization = jwt ? `Bearer ${jwt}` : '';

    return { headers: { ...headers, authorization } };
  });

  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: authLink.concat(httpLink),
    cache: apolloCache,
  });
}

export function initializeApollo(
  initialState = null,
  session?: Session | null,
) {
  const apolloClientGlobal = apolloClient ?? createApolloClient(session);

  if (initialState) {
    apolloClientGlobal.cache.restore(
      initialState as NormalizedCacheObject & NormalizedCache,
    );
  }

  if (typeof window === 'undefined') return apolloClientGlobal;

  apolloClient = apolloClient ?? apolloClientGlobal;

  return apolloClient;
}

export function useApollo(initialState = null, session?: Session) {
  const store = useMemo(() => initializeApollo(initialState, session), [
    initialState,
    session,
  ]);

  return store;
}
