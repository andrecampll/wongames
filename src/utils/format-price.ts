export function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('en', {
    style: 'currency',
    currency: 'USD',
  }).format(price);
}
