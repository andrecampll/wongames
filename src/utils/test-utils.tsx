import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';

import {
  WishlistContext,
  WishlistContextDefaultValues,
  WishlistContextDTO,
} from '../hooks/wishlist/useWishlist';

import theme from '../styles/theme';
import {
  CartContext,
  CartContextDTO,
  CartContextDefaultValues,
} from '../hooks/cart/useCart';

type CustomRenderProps = {
  cartProviderProps?: CartContextDTO;
  wishlistProviderProps?: WishlistContextDTO;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = CartContextDefaultValues,
    wishlistProviderProps = WishlistContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {},
) =>
  render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>
        <WishlistContext.Provider value={wishlistProviderProps}>
          {ui}
        </WishlistContext.Provider>
      </CartContext.Provider>
    </ThemeProvider>,
    renderOptions,
  );

export * from '@testing-library/react';
export { customRender as render };
