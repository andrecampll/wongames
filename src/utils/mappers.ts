import { QueryGames_games } from '../graphql/generated/QueryGames';
import {
  QueryHome_banners,
  QueryHome_sections_freeGames_highlight,
} from '../graphql/generated/QueryHome';
import { QueryWishlist_wishlists_games } from '../graphql/generated/QueryWishlist';

import { formatPrice } from './format-price';

export const bannerMapper = (banners: QueryHome_banners[]) => {
  return banners.map(({ image, title, subtitle, button, ribbon }) => ({
    image: `http://localhost:1337${image?.url}`,
    title,
    subtitle,
    buttonLabel: button?.label,
    buttonLink: button?.link,
    ...(ribbon && {
      ribbon: ribbon.text,
      ribbonColor: ribbon.color,
      ribbonSize: ribbon.size,
    }),
  }));
};

export const gamesMapper = (games: QueryGames_games[] | null | undefined) => {
  return games
    ? games.map(({ id, name, slug, developers, cover, price }) => ({
        id,
        title: name,
        slug,
        developer: developers[0].name,
        image: cover?.url
          ? `http://localhost:1337${cover?.url}`
          : 'https://i.stack.imgur.com/y9DpT.jpg',
        price,
      }))
    : [];
};

export const highlightMapper = (
  highlight: QueryHome_sections_freeGames_highlight | null | undefined,
) => {
  return highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: `http://localhost:1337${highlight.background.url}`,
        floatImage: `http://localhost:1337${highlight.floatImage.url}`,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink,
        alignment: highlight.alignment,
      }
    : {};
};

export const cartMapper = (
  games:
    | QueryGames_games[]
    | QueryWishlist_wishlists_games[]
    | null
    | undefined,
) => {
  return games
    ? games.map(game => ({
        id: game.id,
        img: `http://localhost:1337${game.cover?.url}`,
        title: game.name,
        price: formatPrice(game.price),
      }))
    : [];
};
