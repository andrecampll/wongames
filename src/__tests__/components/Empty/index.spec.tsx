import { screen } from '@testing-library/react';
import { render } from '../../../utils/test-utils';
import Empty, { EmptyProps } from '../../../components/Empty';

const props: EmptyProps = {
  title: 'Your wishlist is empty',
  description: 'Simple description',
};

describe('Empty', () => {
  it('should be able to render', () => {
    const { container } = render(<Empty {...props} />);

    expect(
      screen.getByRole('image', {
        name: /a gamer in a couch playing videogame/i,
      }),
    ).toBeInTheDocument();

    expect(screen.getByRole('heading', { name: /your wishlist is empty/i }));
    expect(screen.getByText(/simple description/i)).toBeInTheDocument();

    expect(
      screen.getByRole('link', { name: /go back to store/i }),
    ).toHaveAttribute('href', '/');

    expect(container.parentElement).toMatchSnapshot();
  });

  it('should not render the link hasLink is false', () => {
    render(<Empty {...props} hasLink={false} />);

    expect(
      screen.queryByRole('link', { name: /go back to store/i }),
    ).not.toBeInTheDocument();
  });
});
