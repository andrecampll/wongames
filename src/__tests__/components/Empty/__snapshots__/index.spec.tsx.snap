// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty should be able to render 1`] = `
<body>
  <div>
    <div
      class="sc-gsTCUz lapHmA"
    >
      <img
        alt="a gamer in a couch playing videogame"
        class="sc-dlfnbm jmjOSl"
        role="image"
        src="/img/empty.svg"
      />
      <h2
        class="sc-hKgILt grkVlg"
      >
        Your wishlist is empty
      </h2>
      <p
        class="sc-eCssSg etWsxR"
      >
        Simple description
      </p>
      <a
        class="sc-bdfBwQ gGQfXn"
        href="/"
      >
        <span>
          Go back to store
        </span>
      </a>
    </div>
  </div>
</body>
`;
